cmd/solana_exporter/slots.go

```diff
@@ -75,112 +90,141 @@ func (c *solanaCollector) WatchSlots() {
+				"counters updated to slot %d (+%d), epoch %d (slots %d-%d, %d remaining)",
+				last,
+				last-watermark,
+				currentEpoch,
+				firstSlot,
+				lastSlot,
+				lastSlot-last,
+			)
+
+			watermark = last
+			currentEpoch, firstSlot, lastSlot = getEpochBounds(info)
+
+			currentEpochNumber.Set(float64(currentEpoch))
+			epochFirstSlot.Set(float64(firstSlot))
+			epochLastSlot.Set(float64(lastSlot))
 		}
 
-		klog.Infof("confirmed slot %d (offset %d, +%d), epoch %d (from slot %d to %d, %d remaining)",
-			info.AbsoluteSlot, info.SlotIndex, info.SlotIndex-watermark, info.Epoch, firstSlot, lastSlot, lastSlot-info.AbsoluteSlot)
-
-		// Get list of confirmed blocks since the last request. This is totally undocumented, but the result won't
-		// contain missed blocks, allowing us to figure out block production success rate.
-		rangeStart := firstSlot + watermark
-		rangeEnd := firstSlot + info.SlotIndex - 1
+		totalTransactionsTotal.Set(float64(info.TransactionCount))
+		confirmedSlotHeight.Set(float64(info.AbsoluteSlot))
 
-		ctx, cancel = context.WithTimeout(context.Background(), httpTimeout)
-		cfm, err := c.rpcClient.GetConfirmedBlocks(ctx, rangeStart, rangeEnd)
+		last, err := updateCounters(c.rpcClient, currentEpoch, watermark, nil)
 		if err != nil {
-			klog.Errorf("failed to request confirmed blocks at %d, retrying: %v", watermark, err)
-			cancel()
+			klog.Info(err)
 			continue
 		}
-		cancel()
 
-		klog.V(1).Infof("confirmed blocks: %d -> %d: %v", rangeStart, rangeEnd, cfm)
+		klog.Infof(
+			"counters updated to slot %d (offset %d, +%d), epoch %d (slots %d-%d, %d remaining)",
+			last,
+			info.SlotIndex,
+			last-watermark,
+			currentEpoch,
+			firstSlot,
+			lastSlot,
+			lastSlot-last,
+		)
+
+		watermark = last
+	}
+}
 
-		// Figure out leaders for each block in range
-		for i := watermark; i < info.SlotIndex; i++ {
-			leader, ok := epochSlots[i]
-			abs := firstSlot + i
-			if !ok {
-				// This cannot happen with a well-behaved node and is a programming error in either Solana or the exporter.
-				klog.Fatalf("slot %d (offset %d) missing from epoch %d leader schedule",
-					abs, i, info.Epoch)
-			}
+// getEpochBounds returns the epoch, first slot and last slot given an EpochInfo struct
+func getEpochBounds(info *rpc.EpochInfo) (int64, int64, int64) {
+	firstSlot := info.AbsoluteSlot - info.SlotIndex
+	lastSlot := firstSlot + info.SlotsInEpoch
 
-			// Check if block was included in getConfirmedBlocks output, otherwise, it was skipped.
-			var present bool
-			for _, s := range cfm {
-				if abs == s {
-					present = true
-				}
-			}
+	return info.Epoch, firstSlot, lastSlot
+}
 
-			var skipped string
-			var label string
-			if present {
-				skipped = "(valid)"
-				label = "valid"
-			} else {
-				skipped = "(SKIPPED)"
-				label = "skipped"
-			}
+func updateCounters(c *rpc.RPCClient, epoch, firstSlot int64, lastSlotOpt *int64) (int64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
```

Comment:
`go vet` seems happy with this, but you're never actually canceling this context with something like a defer. Is that on purpose?

If not, then I believe it is going to [hold on to those resources](https://github.com/golang/go/blob/1831437f197d43608c8086dc42530a89d975aba2/src/context/context.go#L20-L23) for the lifetime of this process.

Reply:
I'm calling `cancel()` after every RPC call, I started doing that in the main loop (since it never returns therefore `defer` will never be called) and it stayed that way after refactoring into `updateCounters`, but yes, `updateCounters` should be using `defer`.